common/lib/xmodule/xmodule/vertical_block.py

```diff
@@ -27,10 +28,27 @@
 # OBSOLETE: This obsoletes 'type'
 CLASS_PRIORITY = ['video', 'problem']
 
+# Make '_' a no-op so we can scrape strings. Using lambda instead of
+#  `django.utils.translation.ugettext_noop` because Django cannot be imported in this file
+_ = lambda text: text
+
+
+class VerticalFields(object):
```

Comment:
I wonder if we should be annotating _all_ block types as explicitly comment-able, instead of special-casing (and recursively traversing) verticals? Or maybe just the non-leaf nodes (course, sequence, vertical, etc...)

Right now, we're sort of implicitly adding support at the sequential/chapter/course level, but only in that (if I'm reading this correctly) we derive the support from all of the verticals; ie: `course['discussion_enabled']` doesn't _really_ mean it's enabled for that course, but rather that it is enabled for _all_ _verticals_ w/in the course. Maybe the semantics don't matter too much, but it's a little confusing now and I suspect it will make things more confusing later...

Anyway, it seems like a lot of work to have to traverse the course structure to compute values each time we want to see if its enabled. Worse, I suspect there may be cases where we're probably walking the course (and/or parts of it) multiple times per request (once for course, then for chapters, etc...), though I didn't track the stack trace deeply.

I'm admittedly not too up-to-date on the inner workings of Studio publishing, but it seems like there should be a way for us to either, memoize or attach directly, the enabled flag on any/all applicable blocks at Publish time, instead of recomputing on each lookup.

Reply:
The idea here is that  `discussion_enabled` at the course level is a global on-off switch for discussions for that course. It is what is set when you click the "Enable" button for Discussions in the new course authoring MFE. 

The original idea was to add this field to all Blocks, but it was getting a but confusing to be honest. What does it mean if a section is commentable vs a unit is commentable? Where will the UI show up in that case and should the section's discussion embed include all discussions under that section or have a separate section-level discussion? So, I kept it restrained to just units. 

All this data will be pushed to models on publish time. I have another in the backlog to add this to block transformers and to the new learning sequences/outlines API. 

I think we can potentially just remove the display of the current state at the non-unit level. We can add a visual indicator, like an icon in the studio UI that signifies which units are discussable, and authors can just get a rough idea looking at that instead of seeing it in a UI.

Reply:
> The idea here is that `discussion_enabled` at the course level is a global on-off switch for discussions for that course. It is what is set when you click the "Enable" button for Discussions in the new course authoring MFE.

Cool, if I'm understanding correctly, is the idea that you'd toggle `discussion_enabled` at the course level to globally enable the feature, which would have the effect of toggling `discussion_enabled` for each unit in the course? And then presumably you could also disable discussions for all units by toggling it _off_ on the course level again (subsequently turning it off for all units)? And also then be able to toggle on/off on a per-unit level too?

That generally makes sense (the recursive setting of the value), though we may want to ensure that, at least on the UI side of things, we draw a distinction between the act of "enabling the _feature_ course-wide" versus "enabling the _integration_ on a per-unit basis". 

I'm concerned about some edge cases, laid out below...

Assuming I read that correctly (please correct me if I'm wrong), I'm a little worried about tying the course-level-feature-toggle to the unit-level-integration-toggle as conflating the values could complicate some use cases. For example, image one were to:
- enable discussions course-wide
- toggle _off_ discussions for some units
- decide I want to temporarily _disable_ discussion course-wide, either for a time period (not necessarily "blackout dates" like the current UI, though it could be, also holidays, etc), or some indeterminate period of time (maybe the forums are being spammed, or you just don't like the experience and want to try without, etc.).
- later decide I want to _re_ _enable_ discussions
- now I have to remember to repeat step 2 from above (toggle it _off_ for certain units again) or discussions will start showing up on all the units where I had previously disabled it.

*cue angry instructors and course designers*

There will also be courses that want to use the new experience, but only on a small percent of their units; how will we enable that? With the current approach, it seems like we'd have to:
- enable it course-wide
- go through the course and manually disable it on most units
  - this could be a nightmare on large courses
- pray that you never "mistakenly" disable then re-enable it, falling prey to the use-case above (and needing to repeat step 2 here again, which would be extra painful)

This all makes me think we might want to split this into two toggles:
- course-level: is the feature enabled for this course?
  - disclaimer: I think this concept does exist (via the integration's configuration) and maybe I'm looking at this PR here too narrowly and not seeing how the parts fit together.
- unit-level: is the feature _integrated_ into this unit?
  - this ensure that toggling the feature on-and-off course-wide will not overwrite any per-unit settings.

The idea being, you'd enable/disable on the course level and then presumably there'd be a button or checkbox or some kind of action you can optionally perform to either set or unset it recursively (for all units), either at enable-time or after-the-fact.

So again, maybe this is covered elsewhere, but wanted to flag these cases.

> What does it mean if a section is commentable vs a unit is commentable? Where will the UI show up in that case and should the section's discussion embed include all discussions under that section or have a separate section-level discussion?

Heh, this is what I was wondering too and what prompted my question ;)

I wanted to reserve judgement in case you folks had a use-case I wasn't seeing, but we may be on the same page; I'm equally unclear on on what it would mean to make a section commentable and/or how we'd represent that in the UI (and what actions it would enable, etc.).

So if we don't have a clear use-case here, I'm inclined to drop support.

> All this data will be pushed to models on publish time. I have another in the backlog to add this to block transformers and to the new learning sequences/outlines API.

Ah cool. Where possible, feel free to make a quick note of things in the PR ("this will be optimized later", etc.), if you want to save yourself from (some of) my uninformed questions ;)

> I think we can potentially just remove the display of the current state at the non-unit level. We can add a visual indicator, like an icon in the studio UI that signifies which units are discussable, and authors can just get a rough idea looking at that instead of seeing it in a UI.

Cool, that jibes with what I was thinking. If we find we need more UI granularity later, we can consider sending back more data, but I'm inclined to keep it simple for now :)

Reply:
So the conversation has moved on a bit from when I was tagged, but fwiw:

1. I do think it's fine to add a field to the `VerticalBlock` for this functionality.
2. I agree with @xitij2000's comments later on that we should name it to something that better disambiguates it from configuration settings, and focuses more on how it's describing the content itself. `is_discussable` or `discussable` sound fine to me in that regard.